crates/holochain/src/core/workflow/app_validation_workflow/tests.rs

```diff
@@ -1,41 +1,156 @@
+    )
+    .await;
+    let app = conductor.setup_app("", &[dna_file.clone()]).await.unwrap();
+    let cell_id = app.cells()[0].cell_id().clone();
+
+    let app_validation_workspace = Arc::new(AppValidationWorkspace::new(
+        conductor.get_authored_db(&dna_hash).unwrap().into(),
+        conductor.get_dht_db(&dna_hash).unwrap(),
+        conductor.get_dht_db_cache(&dna_hash).unwrap(),
+        conductor.get_cache_db(&cell_id).await.unwrap(),
+        conductor.keystore(),
+        Arc::new(dna_file.dna_def().clone()),
+    ));
+    // check there are no ops to app validate
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap();
+    assert_eq!(ops_to_validate.len(), 0);
+
+    // create ops to validate
+    let action = Action::Dna(Dna {
+        author: fixt!(AgentPubKey),
+        timestamp: Timestamp::now(),
+        hash: dna_hash.clone(),
+    });
+    let dht_op = DhtOp::RegisterAgentActivity(fixt!(Signature), action.clone());
+    let dht_op_hash = DhtOpHash::with_data_sync(&dht_op);
+    let dht_op_hashed = DhtOpHashed::with_pre_hashed(dht_op, dht_op_hash.clone());
+
+    let dht_op_2 = DhtOp::StoreEntry(fixt!(Signature), fixt!(NewEntryAction), fixt!(Entry));
+    let dht_op_hash_2 = DhtOpHash::with_data_sync(&dht_op_2);
+    let dht_op_hashed_2 = DhtOpHashed::with_pre_hashed(dht_op_2, dht_op_hash_2.clone());
+
+    app_validation_workspace
+        .dht_db
+        .write_async({
+            let dht_op_hash = dht_op_hash.clone();
+            let dht_op_hash_2 = dht_op_hash_2.clone();
+            move |txn| {
+                insert_op(txn, &dht_op_hashed).unwrap();
+                put_validation_limbo(txn, &dht_op_hash, ValidationStage::SysValidated).unwrap();
+                insert_op(txn, &dht_op_hashed_2).unwrap();
+                put_validation_limbo(txn, &dht_op_hash_2, ValidationStage::SysValidated)
+            }
+        })
+        .await
+        .unwrap();
+
+    // check there is one op to validate now
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap();
+    assert_eq!(ops_to_validate.len(), 2);
+
+    // run validation workflow
+    // outcome should be complete - all ops should have been validated
+    let app_validation_result = app_validation_workflow_inner(
+        Arc::new(dna_hash.clone()),
+        app_validation_workspace.clone(),
+        conductor.raw_handle(),
+        &conductor.holochain_p2p().to_dna(dna_hash.clone(), None),
+        conductor
+            .get_or_create_space(&dna_hash)
+            .unwrap()
+            .dht_query_cache,
+    )
+    .await;
+    assert_matches!(app_validation_result, Ok(WorkComplete::Complete));
+
+    // check that previously inserted ops have been validated
+    // and are valid
+    let num_pending_ops: usize = app_validation_workspace
```

Comment:
Nice specific check! Is it also worth checking that `get_ops_to_app_validate` has gone back to 0?

Reply:
Why not. Before I had just that check and then changed it this more specific one. Checking both makes it sounder.